### PR TITLE
feat: add notifications endpoint

### PR DESCRIPTION
**Pull request by alifaaziz**
create notifications endpoint #31 :
- get all notifications by userId from token
- read all notifications
- read a notification
- delete notification

### CODE REVIEW

**Comment by ccrsxx** [comment]
@alifaaziz jangan lupa buat postman requestnya, biar testnya entar gampang.

**Comment by alifaaziz** [comment]
> @alifaaziz jangan lupa buat postman requestnya, biar testnya entar gampang.

done bang

File: src/controllers/notification.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+/** @import {Request, Response} from 'express' */
+/** @import {User} from '@prisma/client'} */
+
+/**
+ * @param {Request} _req
+ * @param {Response} res
+ */
+async function getNotifications(_req, res) {
+  const userId = res.locals.user.id;
+  const notifications = await NotificationService.getNotifications(userId);
```
Comment (ccrsxx) [inline src/controllers/notification.js]: nit: lgsg aja res.local.user.id di parameternya, lebih clean, jangan buat variable baru

File: src/controllers/notification.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  res.status(200).json({ data: notifications });
+}
+
+/**
+ * @param {Request} _req
+ * @param {Response} res
+ */
+async function readAllNotifications(_req, res) {
+  const userId = res.locals.user.id;
+  await NotificationService.readAllNotifications(userId);
```
Comment (ccrsxx) [inline src/controllers/notification.js]: nit: lgsg aja res.local.user.id di parameternya, lebih clean, jangan buat variable baru

File: src/controllers/notification.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+  res.status(200).json({ message: 'All notifications marked as read' });
```
Comment (ccrsxx) [inline src/controllers/notification.js]: grammar better ganti ke "All notifications have been marked as read"

File: src/controllers/notification.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+/**
+ * @param {Request} req
+ * @param {Response} res
+ */
+async function readNotification(req, res) {
+  const { id } = req.params;
+  const userId = res.locals.user.id;
+  await NotificationService.readNotification(id, userId);
+
+  res.status(200).json({ message: `Notification ${id} marked as read` });
```
Comment (ccrsxx) [inline src/controllers/notification.js]: ganti ke 'Notification has been marked as read', gk usah mention id

File: src/controllers/notification.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+/**
+ * @param {Request} req
+ * @param {Response} res
+ */
+async function deleteNotification(req, res) {
+  const { id } = req.params;
+  const userId = res.locals.user.id;
+  await NotificationService.deleteNotification(id, userId);
+
+  res.status(200).json({ message: `Notification ${id} deleted` });
```
Comment (ccrsxx) [inline src/controllers/notification.js]: ganti ke  'Notification has been deleted'

File: src/controllers/notification.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+/**
+ * @param {Request} req
+ * @param {Response} res
+ */
+async function deleteNotification(req, res) {
+  const { id } = req.params;
+  const userId = res.locals.user.id;
+  await NotificationService.deleteNotification(id, userId);
```
Comment (ccrsxx) [inline src/controllers/notification.js]: lgsg ke pake param, tanpa buat variable baru

File: src/routes/notifications.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  );
+
+  router.patch(
+    '/read-all',
+    AuthMiddleware.isAuthorized,
+    NotificationController.readAllNotifications
+  );
+
+  router.patch(
+    '/:id',
```
Comment (ccrsxx) [inline src/routes/notifications.js:26]: pake commonValidationUuid sebelum is authorized

File: src/routes/notifications.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    AuthMiddleware.isAuthorized,
+    NotificationController.readNotification
+  );
+
+  router.delete(
+    '/:id',
+    AuthMiddleware.isAuthorized,
```
Comment (ccrsxx) [inline src/routes/notifications.js:36]: pake commonValidationUuid sebelum is authorized

File: src/services/notification.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+ * @param {string} userId
+ */
+async function deleteNotification(notificationId, userId) {
+  const notification = await prisma.notification.findFirst({
+    where: { id: notificationId, userId }
+  });
+
+  if (!notification) {
+    throw new HttpError(404, { message: 'Notification not found' });
+  }
```
Comment (ccrsxx) [inline src/services/notification.js]: ini bisa di buat middleware abis `isAuthorized`, namanya `isNotificationExists`. jadi bisa reusable di dua routes.
jadi bisa remove check notif gk diatas, jadi pure lewat middleware.

buat middleware-nya di `/middlewares/notification.js`.

**Review by ccrsxx — APPROVED** [review]
Overall good, tapi perlu improvement lagi.

**Review by ccrsxx — CHANGES_REQUESTED** [review]
Overall good, tapi perlu improvement lagi.

**Comment by ccrsxx** [comment]
image

btw bad response yg 400 sama 404 belum ada.

**Review by niemoo — APPROVED** [review]
LGTM

**Review by ccrsxx — APPROVED** [review]
OK

**Pull request merged**